[fuzzer] $INCLUDE hanging indefinitely

The fuzzer identified more errors with the `FuzzNewRR` entry point (which is much better than if it was the main `Fuzz` one.

TL;DR scan can hang indefinitely if passed a file like /dev/stdout or stderr

```patch
diff --git a/fuzz_test.go b/fuzz_test.go
index d597f41..742dd7e 100644
--- a/fuzz_test.go
+++ b/fuzz_test.go
@@ -252,6 +252,10 @@ func TestNewRRCommentLengthCrasherString(t *testing.T) {
                                ";;;;;;;;;;;",
                        `dns: bad HINFO Fields: "comment length insufficient for parsing" at line: 1:1951`,
                },
+               {
+                       "$GENERATE 1-99 $$iNCLUDE /proc/self/fd/1 \xcb.",
+                       "a",
+               },
        }

        for _, tc := range tests {
```

or anything with `$INCLUDE /proc/self/fd/1`

Basically, parsing tokens will hang.


Using stdin would crash:
```
$ go test -run TestNewRRCommentLengthCrasherString
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x7395ca]

goroutine 7 [running]:
testing.tRunner.func1(0xc000100300)
        /usr/lib/golang/src/testing/testing.go:792 +0x387
panic(0x7a27c0, 0xacceb0)
        /usr/lib/golang/src/runtime/panic.go:513 +0x1b9
github.com/miekg/dns.TestNewRRCommentLengthCrasherString.func1(0xc000100300)
        /home/chantra/go/src/github.com/miekg/dns/fuzz_test.go:267 +0x6a
testing.tRunner(0xc000100300, 0xc000048d40)
        /usr/lib/golang/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
        /usr/lib/golang/src/testing/testing.go:878 +0x35c
exit status 2
FAIL    github.com/miekg/dns    0.011s
```

Maybe a simple approach would be to enforce a regular file. e.g 

```patch
diff --git a/fuzz_test.go b/fuzz_test.go
index d597f41..b11fcf3 100644
--- a/fuzz_test.go
+++ b/fuzz_test.go
@@ -252,6 +252,14 @@ func TestNewRRCommentLengthCrasherString(t *testing.T) {
                                ";;;;;;;;;;;",
                        `dns: bad HINFO Fields: "comment length insufficient for parsing" at line: 1:1951`,
                },
+               {
+                       "$GENERATE 1-99 $$iNCLUDE /proc/self/fd/2 \xcb.",
+                       "dns: not a regular file `/proc/self/fd/2': <nil>: \"/proc/self/fd/2\" at line: 1:25",
+               },
+               {
+                       "$iNCLUDE /proc/self/fd/0",
+                       "dns: not a regular file `/proc/self/fd/0': <nil>: \"/proc/self/fd/0\" at line: 1:24",
+               },
        }

        for _, tc := range tests {
diff --git a/scan.go b/scan.go
index 0f9361a..df08c2f 100644
--- a/scan.go
+++ b/scan.go
@@ -489,6 +489,26 @@ func (zp *ZoneParser) Next() (RR, bool) {
                                msg := fmt.Sprintf("failed to open `%s'%s: %v", l.token, as, e1)
                                return zp.setParseError(msg, l)
                        }
+                       s1, e1 := r1.Stat()
+                       if e1 != nil {
+                               var as string
+                               if !filepath.IsAbs(l.token) {
+                                       as = fmt.Sprintf(" as `%s'", includePath)
+                               }
+
+                               msg := fmt.Sprintf("failed to get stats `%s'%s: %v", l.token, as, e1)
+                               return zp.setParseError(msg, l)
+                       }
+
+                       if !s1.Mode().IsRegular() {
+                               var as string
+                               if !filepath.IsAbs(l.token) {
+                                       as = fmt.Sprintf(" as `%s'", includePath)
+                               }
+
+                               msg := fmt.Sprintf("not a regular file `%s'%s: %v", l.token, as, e1)
+                               return zp.setParseError(msg, l)
+                       }

                        zp.sub = NewZoneParser(r1, neworigin, includePath)
                        zp.sub.defttl, zp.sub.includeDepth, zp.sub.osFile = zp.defttl, zp.includeDepth+1, r1
```

The above test would only work on Linux, replacing with /dev/stdin, /dev/std{out,err} should do BSDs + Linux

--- a/fuzz.go
+++ b/fuzz.go
@@ -2,6 +2,8 @@
 
 package dns
 
+import "strings"
+
 func Fuzz(data []byte) int {
 	msg := new(Msg)
 
@@ -16,7 +18,14 @@ func Fuzz(data []byte) int {
 }
 
 func FuzzNewRR(data []byte) int {
-	if _, err := NewRR(string(data)); err != nil {
+	str := string(data)
+	// Do not fuzz lines that include the $INCLUDE keyword and hint the fuzzer
+	// at avoiding them.
+	// See GH#1025 for context.
+	if strings.Contains(strings.ToUpper(str), "$INCLUDE") {
+		return -1
+	}
+	if _, err := NewRR(str); err != nil {
 		return 0
 	}
 	return 1

